operator/internal/controller/pulumi/git.go

```diff
@@ -18,318 +18,240 @@ import (
+		return nil, fmt.Errorf(`missing "commit" or "branch"`)
+	}
+	if gs.Commit != "" && gs.Branch != "" {
+		return nil, fmt.Errorf(`only one of "commit" or "branch" should be specified`)
+	}
 
-	// 		if sess.stack.GitAuth.PersonalAccessToken != nil {
-	// 			accessToken, err := sess.resolveResourceRef(ctx, sess.stack.GitAuth.PersonalAccessToken)
-	// 			if err != nil {
-	// 				return nil, fmt.Errorf("resolving gitAuth personal access token: %w", err)
-	// 			}
-	// 			gitAuth.PersonalAccessToken = accessToken
-	// 			return gitAuth, nil
-	// 		}
+	ref := gs.Commit
+	if gs.Branch != "" {
+		ref = gs.Branch
+	}
+	url := gs.ProjectRepo
 
-	// 		if sess.stack.GitAuth.BasicAuth == nil {
-	// 			return nil, errors.New("gitAuth config must specify exactly one of " +
-	// 				"'personalAccessToken', 'sshPrivateKey' or 'basicAuth'")
-	// 		}
+	return newGitSource(url, ref, auth)
+}
 
-	// 		userName, err := sess.resolveResourceRef(ctx, &sess.stack.GitAuth.BasicAuth.UserName)
-	// 		if err != nil {
-	// 			return nil, fmt.Errorf("resolving gitAuth username: %w", err)
-	// 		}
+func newGitSource(rawURL string, ref string, auth *auto.GitAuth) (Source, error) {
+	url, _, err := gitutil.ParseGitRepoURL(rawURL)
+	if err != nil {
+		return nil, fmt.Errorf("parsing %q: %w", rawURL, err)
+	}
+	refSpec := config.RefSpec(fmt.Sprintf("+%s:%s", ref, ref))
+
+	fs := memory.NewStorage()
+	remote := git.NewRemote(fs, &config.RemoteConfig{
+		Name:  "origin",
+		URLs:  []string{url},
+		Fetch: []config.RefSpec{refSpec},
+	})
+
+	return &gitSource{
+		fs:     fs,
+		ref:    refSpec,
+		remote: remote,
+		auth:   auth,
+	}, nil
+}
 
-	// 		password, err := sess.resolveResourceRef(ctx, &sess.stack.GitAuth.BasicAuth.Password)
-	// 		if err != nil {
-	// 			return nil, fmt.Errorf("resolving gitAuth password: %w", err)
-	// 		}
+type gitSource struct {
+	fs     *memory.Storage
+	ref    config.RefSpec
+	remote *git.Remote
+	auth   *auto.GitAuth
+}
 
-	// 		gitAuth.Username = userName
-	// 		gitAuth.Password = password
-	// 	} else if sess.stack.GitAuthSecret != "" {
-	// 		namespacedName := types.NamespacedName{Name: sess.stack.GitAuthSecret, Namespace: sess.namespace}
+func (gs gitSource) CurrentCommit(ctx context.Context) (string, error) {
+	// If our ref is already a commit then use it directly.
+	if gs.ref.IsExactSHA1() {
+		return gs.ref.Src(), nil
+	}
 
-	// 		// Fetch the named secret.
-	// 		secret := &corev1.Secret{}
-	// 		if err := sess.kubeClient.Get(ctx, namespacedName, secret); err != nil {
-	// 			sess.logger.Error(err, "Could not find secret for access to the git repository",
-	// 				"Namespace", sess.namespace, "Stack.GitAuthSecret", sess.stack.GitAuthSecret)
-	// 			return nil, err
-	// 		}
+	// Otherwise fetch the most recent commit for the ref (branch) we care
+	// about.
+	auth, err := gs.authMethod()
+	if err != nil {
+		return "", fmt.Errorf("getting auth methodL: %w", err)
+	}
+	err = gs.remote.FetchContext(ctx, &git.FetchOptions{
```

Comment:
Did it prove impractical to avoid fetching the bits?

Reply:
Updated this to essentially perform an ls-remote and then grep the remote refs. Tests went from ~1.5s to ~1.0s when run against p-examples.